flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -409,6 +416,51 @@ struct CharacterOpsBuilderImpl {
 
     return {substringRef, substringLen};
   }
+
+  mlir::Value createLenTrim(Char str) {
+    // Note: Runtime for LEN_TRIM should also be available at some
+    // point. For now use an inlined implementation.
+    auto indexType = builder.getIndexType();
+    mlir::Value len = builder.createHere<fir::ConvertOp>(indexType, str.len);
+    auto data = materializeValue(str);
+    auto one = builder.createIntegerConstant(indexType, 1);
+    auto i1Type = mlir::IntegerType::get(1, builder.getContext());
+    auto trueVal = builder.createIntegerConstant(i1Type, 1);
+    auto blank = createBlankConstant(str.getCharacterType());
+    // FIXME: Conversion wrong in general, it would be cool to compare fir.char
+    // without cast. Needs the kindmap.
+    auto i8Type = mlir::IntegerType::get(8, builder.getContext());
+    blank = builder.createHere<fir::ConvertOp>(i8Type, blank);
+    // initVal must not be random to account for the zero length case where the
+    // no iteration of the loop is ever executed.
+    mlir::Value initVal = builder.createHere<mlir::SubIOp>(len, one);
+
+    auto iterWhile =
+        builder.createHere<fir::IterWhileOp>(one, len, one, trueVal, initVal);
+    auto insPt = builder.saveInsertionPoint();
+    builder.setInsertionPointToStart(iterWhile.getBody());
+
+    auto index = iterWhile.getInductionVar();
+    auto charPosition = builder.createHere<mlir::SubIOp>(len, index);
```

Comment:
Could there be an off-by-one issue here?  If `index` is a value in `[1..len]`, then it looks like `charPosition` is a value in `[len-1..0]`.  Is that correct?  Shouldn't `charPosition` be a value in `[len..1]`?

Whether or not that is an issue, how about changing the loop control to iterate down over the range `[len..1]` instead, to allow direct use of the current `index` to determine `charPosition`?  Doing that might also allow the value of `index` after the loop to serve as the intrinsic call result, although getting that value right for all cases might take a bit more thought.

Or is this code mostly just a temporary place holder, waiting for more permanent code as suggested by the `Note`, and the details don't matter too much?

Reply:
> Could there be an off-by-one issue here? If index is a value in [1..len], then it looks like charPosition is a value in [len-1..0]. Is that correct? Shouldn't charPosition be a value in [len..1]?

Yes, that is correct, `fir.coordinate_of` is zero based so charPosition needs to be in [len-1..0].
I did not try loop with a negative step, that might simplify things a bit. I'll give it a brief try.

Reply:
I've followed your suggestion.

Reply:
Val, I tested that in a bad way, it appears you cannot pass a negative step to a `fir.iterate_while`, so I went back to the original form.

Reply:
That's a bug. Is the verifier failing?